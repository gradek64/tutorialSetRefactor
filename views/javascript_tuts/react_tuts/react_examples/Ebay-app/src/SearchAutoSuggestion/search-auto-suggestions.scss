.search-auto-suggetions {
  position: relative;

  .custom-search {
    width: 100%;
  }

  ul.autocomplete-items {
    position: absolute;
    width: 100%;
    z-index: 99;
    list-style: none;
    box-shadow: 0 10px 6px -6px #777;
    padding: 0;
    margin: 0;
    background-color: white;
    box-shadow: rgba(0, 0, 6, 0.8);

    & {
      border: 1px solid;
    }

    li {
      padding: 0.1rem;

      &:hover {
        background-color: DodgerBlue !important;
        color: #ffffff;
      }

      &.autocomplete-active {
        /*when navigating through the items using the arrow keys:*/
        background-color: DodgerBlue !important;
        color: #ffffff;
      }
    }
  }
}