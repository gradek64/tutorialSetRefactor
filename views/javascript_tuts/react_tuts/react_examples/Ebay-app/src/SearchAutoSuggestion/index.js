import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

// custom;
import SearchBox from '../functional/SearchBox';
import './search-auto-suggestions.scss';
/*
  *@ToggleState is class componet that holds only logic and uses
  *@props.render method to exposed its renderMethod and toglle state
  *@SimpleButton is extending ToggleState (using its logic) and then whole thing (ToggleState and Simple button)
  *@is wrapped in ToggleSimpleButton functional component that is expoerted outside and takes props for SingleButton
*/
class SearchAutoSuggestion extends React.Component {
  /* eslint-disable no-lone-blocks */

  state = {
    suggestions: null,
  }

  suggestionsListRef = React.createRef();

  currentItemIndex = 0

  resetList = () => {
    console.log('cleaing');
    this.setState({ suggestions: null });
    this.currentItemIndex = 0;
  }

  componentDidMount = () => {
    document.addEventListener('click', this.resetList);
  }

  componentWillUnmount = () => {
    document.removeEventListener('click', this.resetList);
  }

  removeActiveClass = (e) => {
    const suggestionItems = this.suggestionsListRef.current.getElementsByTagName('li');
    Array.from(suggestionItems).forEach(node => node.classList.remove('autocomplete-active'));
    if (e) {
      const { searchBoxRef } = this.props;
      searchBoxRef.current.value = e.target.innerHTML;
    }
  }

  /* eslint-enable */
  arrowDownUpHandler = (e) => {
    const suggestionItems = this.suggestionsListRef.current
      ? this.suggestionsListRef.current.getElementsByTagName('li') : {};
    const { searchBoxRef } = this.props;
    // on arrow up;
    if (e.key === 'ArrowUp') {
      // remove previous;
      this.removeActiveClass();
      this.currentItemIndex > 0 ? this.currentItemIndex -= 1 : this.currentItemIndex;
      if (suggestionItems[this.currentItemIndex]) {
        suggestionItems[this.currentItemIndex].classList.add('autocomplete-active');
        searchBoxRef.current.value = suggestionItems[this.currentItemIndex].innerHTML;
      }
    }
    // on arrow down;
    if (e.key === 'ArrowDown') {
      // remove previous;
      this.removeActiveClass();
      if (suggestionItems[this.currentItemIndex]) {
        suggestionItems[this.currentItemIndex].classList.add('autocomplete-active');
        searchBoxRef.current.value = suggestionItems[this.currentItemIndex].innerHTML;
      }
      this.currentItemIndex < suggestionItems.length - 1 ? this.currentItemIndex += 1 : this.currentItemIndex;
    }

    if (e.key === 'Enter' || e.key === 'Escape') {
      this.setState({ suggestions: null });
      this.currentItemIndex = 0;
    }
  }


  onInputHandler = (e) => {
    const triggerSearchPhrase = e.target.value;
    // if it is empty then remove the suggestions
    if (triggerSearchPhrase === '') { this.resetList(); return; }
    /* axios
      .get('http://autosug.ebaystatic.com/autosug',
        {
          headers: { 'Access-Control-Allow-Origin': '*' },
          params:
        {
          kwd: triggerSearchPhrase,
          _jgr: 1,
          sId: 3,
          _ch: 0,
          callback: 'showResults',
        },
        })
      .then((res) => {
        const { data } = res;
        const [, result] = data.match(/showResults\((.*)\)/);
        const callBackRes = JSON.parse(result);
        const suggestions = callBackRes.res ? callBackRes.res.sug : null;
        this.setState({ suggestions });
      }); */
    // harcoded value for search
    const suggestions = ['bike', 'bike lock', 'bike lights',
      'bike pump', 'mountain bike', 'bike trailer', 'bike rack',
      'bike cover', 'electric bike', 'bike bell'];
    this.setState({ suggestions });
  }


  /* toggleMethod is exposed from ToggleState class and allows to as well as toggle state
true or false , that could be used in other componenst as Simple button */

  render() {
    const { submitHandler, searchBoxRef, classSent } = this.props;
    const { suggestions } = this.state;
    return (

      <React.Fragment>
        {/* Toggle State logic is not needed since we show and hide suggestions based on input characters */}
        {/* <ToggleState
        renderComp={(toggleMethod, toggle) => ( */}
        {/* initial thought to have overlay to register click but prevernts any other action on the page */}
        {/* eslint-disable max-len */}
        {/* suggestions ? <ClickAwayListener clickAway={() => { this.setState({ suggestions: null }); }} /> : null */}
        <form className={`search-auto-suggetions ${classSent}`} autoComplete="off">
          <SearchBox
            id="results"
            className="custom-search"
            submitHandler={submitHandler}
            onFucusHandler={this.resetList}
            onInputHandler={this.onInputHandler}
            arrowDownUpHandler={this.arrowDownUpHandler}
            ref={searchBoxRef}
          />
          {suggestions ? (
            <ul
              role="presentation"
              // required by airbnb but div in this case can not be in focus
              onKeyPress={() => {}}
              className="autocomplete-items"
              ref={this.suggestionsListRef}
            >
              {suggestions.map((value, i) => (
                <li
                  key={`suggestions${i}`}
                  role="presentation"
                  className="list-item"
                  onMouseEnter={this.removeActiveClass}
                  onClick={() => { submitHandler(); this.resetList(); }}
                >
                  {value}
                </li>))}

            </ul>) : null}
        </form>
      </React.Fragment>
    );
  }
}

SearchAutoSuggestion.propTypes = {
  classSent: PropTypes.string,
  submitHandler: PropTypes.func.isRequired,
  searchBoxRef: PropTypes.instanceOf(Object).isRequired,
};
SearchAutoSuggestion.defaultProps = {
  classSent: '',
};
export default SearchAutoSuggestion;
