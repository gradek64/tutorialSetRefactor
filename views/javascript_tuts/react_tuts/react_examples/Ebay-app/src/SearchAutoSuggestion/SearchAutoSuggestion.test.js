import React from 'react';
import { mount } from 'enzyme';
// import axios from 'axios';

import SearchAutoSuggestion from './index';

const props = {
  submitHandler: () => {},
  searchBoxRef: React.createRef(),
};
const wrapper = mount(<SearchAutoSuggestion {...props} />);

test('renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('renders suggestions list once state is fed', () => {
  expect(wrapper.find('ul')).toHaveLength(0);
  const suggestions = ['one', 'two'];
  wrapper.setState({ suggestions });
  expect(wrapper.find('ul.autocomplete-items')).toHaveLength(1);
});

test('onInput change calls axios and feeds suggestions state', () => {
  const wrapper0 = mount(<SearchAutoSuggestion {...props} />);

  // simulate onInput event
  const actualInput = wrapper0.find('input#results');
  // jest.mock('axios');
  // simulate input event
  // actualInput.simulate('input', { target: { value: 'Greg' } });
  wrapper0.instance().onInputHandler({ target: { value: 'Greg' } });
  // axios.get.mockResolvedValue(resp);
  /* axios.get = jest.fn(x => Promise.resolve(x));

  axios.get(2).then((data) => {
    expect(data).toBe(2);
  }); */

  // this only will work if you execute this function() it wont detect when event id dispached or any event callback
  // expect(wrapper1.instance().onInputHandler).toHaveBeenCalled();
});

test('it changes the serach box value on arrow down', () => {
  wrapper.setState({ suggestions: ['one', 'two'] });
  // very useful to see the structure of the wrapper component
  // console.log(wrapper.debug());

  // input value is kepth in React.ref();
  const serachBox = wrapper.props().searchBoxRef.current;
  // press arrow donw key on input itself not ref;
  const actualInput = wrapper.find('input#results');
  actualInput.simulate('keydown', { key: 'ArrowDown' });
  expect(serachBox.value).toBe('one');

  // one more key down
  actualInput.simulate('keydown', { key: 'ArrowDown' });
  expect(serachBox.value).toBe('two');
});

test('it changes the serach box value on arrow down and clear suggestions on Enter key down', () => {
  /*
    this test goes into steps so it is sort of small intergration test
    it would be possible to achieve this breaking into smaller tests outside context
    but it wont be much or readiblity so this intgragtion is better approach
  */


  // wrapper needs to be reset to basic one not affected but previous tests;
  const wrapper1 = mount(<SearchAutoSuggestion {...props} />);
  wrapper1.setState({ suggestions: ['one', 'two'] });
  // input value is kepth in React.ref();
  const serachBox = wrapper1.props().searchBoxRef.current;
  // press arrow donw key on input itself not ref;
  const actualInput = wrapper1.find('input#results');
  actualInput.simulate('keydown', { key: 'ArrowDown' });
  expect(serachBox.value).toBe('one');

  // on Enter key down clears the suggestions list
  actualInput.simulate('keydown', { key: 'Enter' });
  expect(wrapper1.find('ul.autocomplete-items')).toHaveLength(0);
});

test('it changes the serach box value on arrow up and clear suggestions on Enter key down', () => {
  // wrapper needs to be reset to basic one not affected but previous tests;
  const wrapper1 = mount(<SearchAutoSuggestion {...props} />);
  wrapper1.setState({ suggestions: ['one', 'two'] });
  // input value is kepth in React.ref();
  const serachBox = wrapper1.props().searchBoxRef.current;
  // press arrow donw key on input itself not ref;
  const actualInput = wrapper1.find('input#results');
  actualInput.simulate('keydown', { key: 'ArrowUp' });
  expect(serachBox.value).toBe('one');

  // on Enter key down clears the suggestions list
  actualInput.simulate('keydown', { key: 'Enter' });
  expect(wrapper1.find('ulautocomplete-items')).toHaveLength(0);
});

test('it changes the serach box value on mouse enter', () => {
  // wrapper needs to be reset to basic one not affected but previous tests;
  const wrapper2 = mount(<SearchAutoSuggestion {...props} />);
  wrapper2.setState({ suggestions: ['one', 'two'] });
  // input value is kepth in React.ref();
  const serachBox = wrapper2.props().searchBoxRef.current;
  // suggestion list
  const suggestionsList = wrapper.find('ul.autocomplete-items');
  const firstItemList = suggestionsList.find('li').at(1);

  // on mouseEnter with event object any has to be defined' on second element
  firstItemList.simulate('mouseenter', { b: 'any' });
  expect(serachBox.value).toBe('two');
});

test('it clears the list on any document element click', () => {
  // wrapper needs to be reset to basic one not affected but previous tests;
  const wrapper3 = mount(<SearchAutoSuggestion {...props} />);
  wrapper3.setState({ suggestions: ['one', 'two'] });

  // simulate click on document
  const bodyElement = document.getElementsByTagName('body')[0];
  // use browser click feature
  bodyElement.click();
  expect(wrapper3.find('ulautocomplete-items')).toHaveLength(0);
});
