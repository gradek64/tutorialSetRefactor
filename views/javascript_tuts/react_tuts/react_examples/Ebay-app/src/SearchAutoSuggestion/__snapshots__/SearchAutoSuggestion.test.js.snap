// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<SearchAutoSuggestion
  classSent=""
  searchBoxRef={
    Object {
      "current": <input
        class="custom-search"
        id="results"
        type="text"
      />,
    }
  }
  submitHandler={[Function]}
>
  <div
    className="search-auto-suggetions "
  >
    <ForwardRef
      arrowDownUpHandler={[Function]}
      className="custom-search"
      id="results"
      onInputHandler={[Function]}
      submitHandler={[Function]}
    >
      <input
        className="custom-search"
        id="results"
        onInput={[Function]}
        onKeyDown={[Function]}
        type="text"
      />
    </ForwardRef>
  </div>
</SearchAutoSuggestion>
`;
